pyro/infer/reparam/conjugate.py

```diff
@@ -88,3 +88,34 @@ def __call__(self, name, fn, obs):
         # Return an importance-weighted point estimate.
         new_fn = dist.Delta(value, log_density=log_density, event_dim=fn.event_dim)
         return new_fn, value
+
+
+class ConjugateLatentReparam(Reparam):
+    """
+    Reparameterize to add an extra ``pyro.deterministic`` site for the latent
+    variable of a conjugate pair distribution.
+
+    This only works for a small set of conjugate pair distributions, including
+    :class:`~pyro.distributions.BetaBinomial` ,
+    :class:`~pyro.distributions.DirichletMultinomial` ,
+    :class:`~pyro.distributions.GammaPoisson` , and
+    :class:`~pyro.distributions.GaussianHMM` .
+    """
+    def __call__(self, name, fn, obs):
+        assert obs is None, "ConjugateLatentReparam does not support observe statements"
+
+        # Sample the observed value.
+        value = pyro.sample("{}_obs".format(name), fn)
+
+        # Expose the latent sample.
+        latent = getattr(value, "_pyro_latent", None)
+        if latent is not None:  # Try to use the latent cached by fn.sample().
+            pyro.deterministic("{}_latent".format(name), latent,
+                               event_dim=fn.event_dim + latent.dim() - value.dim())
+        else:  # Fall back to constucting and sampling from the posterior.
+            latent_fn = fn.get_posterior(value)
+            latent = pyro.sample("{}_latent".format(name), latent_fn)
```

Comment:
@neerajprad wdyt of this conjugacy pattern together with the `._pyro_latent` logic elsewhere in this PR? Note it's rather brittle, e.g. the [torch.stack() logic in Predictive](https://github.com/pyro-ppl/pyro/blob/6e967362ada8ca1661561826656d5cd6ad296ec4/pyro/infer/predictive.py#L47) would need to be applied *after* `ConjugateLatentParam`.

Reply:
I just had a couple of questions to help me understand the changes involved:
 - where is the `get_posterior` method defined?
 - how does this relate to `ConjugateReparam` - are these complimentary approaches (i.e. this is only used for compound distributions) or do we use one for inference and the other for predictions / getting posterior samples?

Reply:
- I have not yet defined `.get_posterior()`, actually I'm still wavering between `.rsample_posterior()` which returns a value, and `.get_posterior()` which returns a distribution (like your `.posterior()` in pyro.contrib.conjugate)
- This PR renames `ConjugateReparam` to `ConjugateUpdateReparam`. The two should be complementary in that `ConjugateUpdateReparam` maps an already-factorized (prior,likelihood) pair of distributions and produces a posterior, whereas `ConjugateLatentReparam` takes a bundled prior+likelihood martinal distribution and pulls out the implicit latent variable. In this PR I am stacking z->y->x and using `ConjugateUpdateReparam` on `p(y|x)` and `ConjugateLatentReparam` on `p(z|y)`.

Note I feel these kludges all point towards a need for Funsors or a move towards Funsors.

Reply:
Interesting. I guess one way to implement the plated observation feature would be to define
```py
posterior = prior.conjugate_update(
    Likelihood.conjugate(data).conjugate_reduce(dim=...))
```
where
- `.conjugate_update()` is like `Funsor.__add__`,
- `.conjugate_reduce(dim)` is like `Funsor.reduce(ops.add, name)` in https://github.com/pytorch/pytorch/pull/32177, and
- `Likelihood.conjugate` is a static method of each class `Likelihood` (e.g. `Poisson`) that produces a corresponding `Gamma`
One issue I'm facing is that `.conjugate(data)` is not always a proper distribution even when `prior + L.conjugate(data)` is proper, e.g. `Poisson.conjugate(0) = Gamma(0, 1)`. I'm tempted to allow improper distributions like `Gamma(0,1)` as we would in Funsor, i.e.
```py
Gamma(0,1, value="z") = Poisson("z", value=0) + const.  # true in Funsor
```
and we'd probably disallow `.sample()` from improper distributions.